### PR TITLE
Feat/sync support2.x#mutiple thread sync02

### PR DESCRIPTION
**Pull request by chenhao26-nineteen**
Feat/sync support2.x#mutiple thread sync02
1.多线程同步nacos to nacos，缩减耗时（当前是200个分组线程处理任务）
2.中心集群
具体实现方案：
https://www.yuque.com/docs/share/7153b166-2570-4708-a7eb-020fb40dfec5?#

### CODE REVIEW

File: nacossync-worker/src/main/java/com/alibaba/nacossync/extension/impl/NacosSyncToNacosServiceImpl.java [hunk 1]
```diff
@@ -39,59 +59,378 @@
...
+        //中心集群，只要不是目标集群传过来的实例，都需要同步（扩展功能）
+        if (!destClusterId.equals(sourceMetaData.get(SOURCE_CLUSTERID_KEY))) {
+            return true;
+        }
+        return false;
+    }
+    
+    private void recordNamingService(TaskDO taskDO, NamingService destNamingService) {
+        String key = taskDO.getId() + ":" + taskDO.getServiceName();
+        Set<NamingService> namingServices = serviceClient.get(key);
```
Comment (paderlol) [inline nacossync-worker/src/main/java/com/alibaba/nacossync/extension/impl/NacosSyncToNacosServiceImpl.java]: 1、please use the  method`computeIfAbsent`  to replace
```
Set<NamingService> namingServices = serviceClient.get(key);
        if(namingServices==null){
            namingServices=new HashSet<>();
        }
        // save dest NamingService
        namingServices.add(destNamingService);
```

2、plz consider using a concurrently safe Set

File: nacossync-worker/src/main/java/com/alibaba/nacossync/extension/impl/NacosSyncToNacosServiceImpl.java [hunk 2]
```diff
@@ -39,59 +59,378 @@
...
-        for (String instanceKey : invalidInstanceKeys) {
-            String[] split = instanceKey.split(":", -1);
-            try {
-                destNamingService.deregisterInstance(taskDO.getServiceName(),
-                        getGroupNameOrDefault(taskDO.getGroupName()), split[0], Integer.parseInt(split[1]));
-            } catch (NacosException e) {
-                log.error("Remove instance={} from Nacos failed", taskDO.getServiceName(), e);
-                
+    private void instanceRemove(List<Instance> destHasSyncInstances, List<Instance> newInstances) {
+        List<Instance> needRemoveInstance = new ArrayList<>();
```
Comment (paderlol) [inline nacossync-worker/src/main/java/com/alibaba/nacossync/extension/impl/NacosSyncToNacosServiceImpl.java:367]: why not just use the `removeAll`
``` java
newInstance.removeAll(destHasSyncInstances);
```

File: nacossync-worker/src/main/java/com/alibaba/nacossync/extension/impl/NacosSyncToNacosServiceImpl.java [hunk 3]
```diff
@@ -102,14 +441,15 @@ private Instance buildSyncInstance(Instance instance, TaskDO taskDO) {
         temp.setHealthy(instance.isHealthy());
         temp.setWeight(instance.getWeight());
         temp.setEphemeral(instance.isEphemeral());
-        Map<String, String> metaData = new HashMap<>(instance.getMetadata());
+        Map<String, String> metaData = new HashMap<>();
```
Comment (paderlol) [inline nacossync-worker/src/main/java/com/alibaba/nacossync/extension/impl/NacosSyncToNacosServiceImpl.java]:
``` 
Map<String, String> metaData = new HashMap<>();
        metaData.putAll(instance.getMetadata());
```
to
```
Map<String, String> metaData = new HashMap<>(instance.getMetadata())
```

File: nacossync-worker/src/main/java/com/alibaba/nacossync/timer/CheckRunningStatusAllThread.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    
+            for (TaskDO taskDO : taskDOS) {
+                List<String> serviceNameList = getServiceNameList(taskDO);
+                if (CollectionUtils.isEmpty(serviceNameList)) {
+                   continue;
+                }
+                
+                //如果是null，证明此时没有处理完成
+                List<String> filterService = serviceNameList.stream()
+                        .filter(s -> skyWalkerCacheServices.getFinishedTask(taskDO.getTaskId() + s ) == null)
```
Comment (paderlol) [inline nacossync-worker/src/main/java/com/alibaba/nacossync/timer/CheckRunningStatusAllThread.java]: The field `s` should have a meaningful name. such as serviceName

File: nacossync-worker/src/main/java/com/alibaba/nacossync/timer/FastSyncHelper.java [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+     * @param filterServices filterServices
+     */
+    public  void syncWithThread(TaskDO taskDO, List<String> filterServices) {
+        long start=System.currentTimeMillis();
+        List<List<String>> lists = averageAssign(filterServices, MAX_THREAD_NUM);
+        CountDownLatch countDownLatch = new CountDownLatch(lists.size());
+        
+        //创建分组线程，每个线程执行自己的任务
+        for (int i = 0; i < lists.size(); i++) {
+            new SyncThread(i, countDownLatch, lists.get(i), taskDO).start();
```
Comment (paderlol) [inline nacossync-worker/src/main/java/com/alibaba/nacossync/timer/FastSyncHelper.java:96]: plz, consider if need to use a thread pool.

File: nacossync-worker/src/main/java/com/alibaba/nacossync/extension/impl/NacosSyncToNacosServiceImpl.java (hunk 2 above)
Comment (chenhao26-nineteen) [reply nacossync-worker/src/main/java/com/alibaba/nacossync/extension/impl/NacosSyncToNacosServiceImpl.java:367]: > 

need to compare the source cluster and the target cluster to find out the instances that need to be de registered in the target cluster, destHasSyncInstances cannot be deleted directly. Here, the parameter name may be ambiguous

File: nacossync-worker/src/main/java/com/alibaba/nacossync/extension/impl/NacosSyncToNacosServiceImpl.java (hunk 1 above)
Comment (chenhao26-nineteen) [reply nacossync-worker/src/main/java/com/alibaba/nacossync/extension/impl/NacosSyncToNacosServiceImpl.java]: > 1、please use the method`computeIfAbsent` to replace
> 
> ```
> Set<NamingService> namingServices = serviceClient.get(key);
>         if(namingServices==null){
>             namingServices=new HashSet<>();
>         }
>         // save dest NamingService
>         namingServices.add(destNamingService);
> ```
> 
> 2、plz consider using a concurrently safe Set

done

File: nacossync-worker/src/main/java/com/alibaba/nacossync/timer/FastSyncHelper.java (hunk 5 above)
Comment (chenhao26-nineteen) [reply nacossync-worker/src/main/java/com/alibaba/nacossync/timer/FastSyncHelper.java:96]: > 

I will consider it

File: nacossync-worker/src/main/java/com/alibaba/nacossync/extension/impl/NacosSyncToNacosServiceImpl.java (hunk 3 above)
Comment (chenhao26-nineteen) [reply nacossync-worker/src/main/java/com/alibaba/nacossync/extension/impl/NacosSyncToNacosServiceImpl.java]: > ```
> Map<String, String> metaData = new HashMap<>();
>         metaData.putAll(instance.getMetadata());
> ```
> 
> to
> 
> ```
> Map<String, String> metaData = new HashMap<>(instance.getMetadata())
> ```

done

File: nacossync-worker/src/main/java/com/alibaba/nacossync/timer/CheckRunningStatusAllThread.java (hunk 4 above)
Comment (chenhao26-nineteen) [reply nacossync-worker/src/main/java/com/alibaba/nacossync/timer/CheckRunningStatusAllThread.java]: > The field `s` should have a meaningful name. such as serviceName

done

**Pull request merged**